### PR TITLE
10.0 sale promotion rule improved lmi

### PR DESCRIPTION
**Pull request by lmignon**
Improve code

### CODE REVIEW

**Comment by lmignon** [comment]
@sebastienbeau Here it's a refactoring of the module to support multi rules and improve performances on rule computation.

File: sale_promotion_rule/README.rst [hunk 1]
```diff
@@ -6,10 +6,20 @@
 Sale Promotion Rule
 ===================
 
-This module add the concept of promotion rule that can be applied on the sale order.
+This module add the concept of promotion rules that can be applied on the sale
```
Comment (rvalyi) [inline sale_promotion_rule/README.rst]: s/add/adds/
s/on the sale order/on a sale order/

File: sale_promotion_rule/README.rst [hunk 2]
```diff
@@ -6,10 +6,20 @@
...
+order.
 
-for now only coupon with discount amount have been implemented.
+Two kinds of rules are implemented:
+ * automatic
+ * coupon
 
+Automatic rules are applied/recomputed automatically for a sale order when the
+user click on the button "Apply discount" into the view form. Depending of the
```
Comment (rvalyi) [inline sale_promotion_rule/README.rst]: s/user click on the button "Apply discount" into the view form/user clicks on the "Apply discount" button in the form view

File: sale_promotion_rule/README.rst [hunk 3]
```diff
@@ -6,10 +6,20 @@
...
+rule's criteria more than one automatic rule can be applied to a same sale
+order.
```
Comment (rvalyi) [inline sale_promotion_rule/README.rst]: s/Depending of/Depending on
s/to a same sale/to a given sale order

File: sale_promotion_rule/README.rst [hunk 4]
```diff
@@ -6,10 +6,20 @@
...
+
+Coupon are special manual rules. Only one coupon can be applied to a
+ale.order. This rule takes always precedence over automatic rules.
```
Comment (rvalyi) [inline sale_promotion_rule/README.rst]: s/ale.order/sale order

**Review by rvalyi — CHANGES_REQUESTED** [review]
some English mistakes...

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 5]
```diff
@@ -106,29 +126,204 @@ def _check_valid_usage(self, order):
...
+        coupon_rule._apply(orders_without_coupon)
+        self.apply_auto(orders_without_coupon)
+
+    @api.model
+    def apply_auto(self, orders):
+        """Apply automatic promotion rules to the orders"""
+        auto_rules = self.search([
+            ('rule_type', '=', 'auto'),
+        ])
+        auto_rules._apply(orders)
```
Comment (bguillot) [inline sale_promotion_rule/models/sale_promotion_rule.py:223]: This crashes if several automatic promotions are found, we should remove the ensure_one() in the _apply() method or make a loop here.

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 6]
```diff
@@ -106,29 +126,204 @@ def _check_valid_usage(self, order):
...
+
+    @api.multi
+    def unset(self, orders):
```
Comment (bguillot) [inline sale_promotion_rule/models/sale_promotion_rule.py]: This method is never called is it usefull ?

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 7]
```diff
@@ -106,29 +126,204 @@ def _check_valid_usage(self, order):
...
+        """
+        Remove the rules from the sale orders and recompute remaining rules.
+        """
+        self.ensure_one()
+        for order in orders:
+            applied_rules = order.promotion_rule_ids + \
+                order.coupon_promotion_rule_id
+            if self not in applied_rules:
+                continue
+            self.reset_promitions()
```
Comment (bguillot) [inline sale_promotion_rule/models/sale_promotion_rule.py]: This method doesn't exists, is it remove_promotions() ?

**Review by bguillot — CHANGES_REQUESTED** [review]
Great improvements ! Thanks !

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 5 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py:223]: oups... The API has changed a lot of time while implementing the new logic. Indeed, we should make a loop here

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 6 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: nor more used... to be removed

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 5 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py:223]: @bguillot code fixed

**Review by bguillot — APPROVED** [review]

**Pull request merged**